consensus/src/dag/mod.rs

```diff
@@ -0,0 +1,641 @@
+use std::collections::HashMap;
+
+use crate::{
+    alerts::{Alert, ForkingNotification},
+    units::{
+        SignedUnit, UncheckedSignedUnit, Unit, UnitStore, Validator as UnitValidator, WrappedUnit,
+    },
+    Data, Hasher, MultiKeychain,
+};
+use log::{debug, trace, warn};
+
+mod reconstruction;
+mod validation;
+
+pub use reconstruction::{ReconstructedUnit, Request};
+use reconstruction::{Reconstruction, ReconstructionResult};
+pub use validation::ValidatorStatus as DagStatus;
+use validation::{Error as ValidationError, Validator};
+
+const LOG_TARGET: &str = "AlephBFT-dag";
+
+/// The result of sending some information to the Dag.
+pub struct DagResult<H: Hasher, D: Data, MK: MultiKeychain> {
+    units: Vec<ReconstructedUnit<SignedUnit<H, D, MK>>>,
+    requests: Vec<Request<H>>,
+    alerts: Vec<Alert<H, D, MK::Signature>>,
+}
+
+impl<H: Hasher, D: Data, MK: MultiKeychain> DagResult<H, D, MK> {
+    fn empty() -> Self {
+        DagResult {
+            units: Vec::new(),
+            requests: Vec::new(),
+            alerts: Vec::new(),
+        }
+    }
+
+    fn alert(alert: Alert<H, D, MK::Signature>) -> Self {
+        DagResult {
+            units: Vec::new(),
+            requests: Vec::new(),
+            alerts: vec![alert],
+        }
+    }
+
+    fn add_alert(&mut self, alert: Alert<H, D, MK::Signature>) {
+        self.alerts.push(alert);
+    }
+
+    fn accumulate(&mut self, other: DagResult<H, D, MK>) {
+        let DagResult {
+            mut units,
+            mut requests,
+            mut alerts,
+        } = other;
+        self.units.append(&mut units);
+        self.requests.append(&mut requests);
+        self.alerts.append(&mut alerts);
+    }
+}
+
+impl<H: Hasher, D: Data, MK: MultiKeychain> From<ReconstructionResult<SignedUnit<H, D, MK>>>
+    for DagResult<H, D, MK>
+{
+    fn from(other: ReconstructionResult<SignedUnit<H, D, MK>>) -> Self {
+        let (units, requests) = other.into();
+        DagResult {
+            units,
+            requests,
+            alerts: Vec::new(),
+        }
+    }
+}
+
+impl<H: Hasher, D: Data, MK: MultiKeychain> From<DagResult<H, D, MK>>
+    for (
+        Vec<ReconstructedUnit<SignedUnit<H, D, MK>>>,
+        Vec<Request<H>>,
+        Vec<Alert<H, D, MK::Signature>>,
+    )
+{
+    fn from(result: DagResult<H, D, MK>) -> Self {
+        let DagResult {
+            units,
+            requests,
+            alerts,
+        } = result;
+        (units, requests, alerts)
+    }
+}
+
+/// The Dag ensuring that all units from the network get returned reconstructed in the correct order.
+pub struct Dag<H: Hasher, D: Data, MK: MultiKeychain> {
+    validator: Validator<H, D, MK>,
+    reconstruction: Reconstruction<SignedUnit<H, D, MK>>,
+}
+
+impl<H: Hasher, D: Data, MK: MultiKeychain> Dag<H, D, MK> {
+    /// A new dag using the provided unit validator under the hood.
+    pub fn new(unit_validator: UnitValidator<MK>) -> Self {
```

Comment:
Nit: could be shortened to 
```rust
Dag {
    validator: Validator::new(unit_validator),
    reconstruction: Reconstruction::new(),
}

Reply:
Fair enough, I dunno why I seem to intuitively prefer first creating the fields. Changed as requested.